crypto/fipsmodule/ec/p384.c

```diff
@@ -503,6 +503,479 @@ static int ec_GFp_nistp384_cmp_x_coordinate(const EC_GROUP *group,
+    out[i] = d;
+    window = (window - d) >> FIAT_P384_SCALAR_RADIX;
+    window += fiat_p384_get_bit(in, (i + 1) * FIAT_P384_SCALAR_RADIX + 1) << 1;
+    window += fiat_p384_get_bit(in, (i + 1) * FIAT_P384_SCALAR_RADIX + 2) << 2;
+    window += fiat_p384_get_bit(in, (i + 1) * FIAT_P384_SCALAR_RADIX + 3) << 3;
+    window += fiat_p384_get_bit(in, (i + 1) * FIAT_P384_SCALAR_RADIX + 4) << 4;
+    window += fiat_p384_get_bit(in, (i + 1) * FIAT_P384_SCALAR_RADIX + 5) << 5;
+  }
+  out[76] = window;
+}
+
+// Compute "textbook" wNAF representation of a scalar.
+// It outputs digits in {0, \pm 1, \pm 3, \pm 5, \pm 7, \pm 9, ...}.
+// A digits is either a zero or an odd integer. It is guaranteed that a non-zero
+// digit is followed by at least (FIAT_P384_SCALAR_DRADIX - 1) zero digits.
+//
+// Note: this function is not constant-time.
+static void fiat_p384_mul_scalar_wnaf(int8_t *out, const unsigned char *in) {
+  int8_t window, d;
+
+  window = in[0] & (FIAT_P384_SCALAR_DRADIX_WNAF - 1);
+  for (size_t i = 0; i < 385; i++) {
+    d = 0;
+    if ((window & 1) && ((d = window & (FIAT_P384_SCALAR_DRADIX_WNAF - 1)) &
+                         FIAT_P384_SCALAR_DRADIX)) {
+      d -= FIAT_P384_SCALAR_DRADIX_WNAF;
+    }
+
+    out[i] = d;
+    window = (window - d) >> 1;
+    window += (fiat_p384_get_bit(in, i + 1 + FIAT_P384_SCALAR_RADIX)) <<
+                                                                      FIAT_P384_SCALAR_RADIX;
+  }
+}
+
+// fiat_p384_select_point selects the |idx|-th projective point from the given
+// precomputed table and copies it to |out| in constant time.
+static void fiat_p384_select_point(fiat_p384_felem out[3],
+                                   size_t idx,
+                                   fiat_p384_felem table[][3],
+                                   size_t table_size) {
+  OPENSSL_memset(out, 0, sizeof(fiat_p384_felem) * 3);
+  for (size_t i = 0; i < table_size; i++) {
+    fiat_p384_limb_t mismatch = i ^ idx;
+    fiat_p384_cmovznz(out[0], mismatch, table[i][0], out[0]);
+    fiat_p384_cmovznz(out[1], mismatch, table[i][1], out[1]);
+    fiat_p384_cmovznz(out[2], mismatch, table[i][2], out[2]);
+  }
+}
+
+// fiat_p384_select_point_affine selects the |idx|-th affine point from
+// the given precomputed table and copies it to |out| in constant-time.
+static void fiat_p384_select_point_affine(fiat_p384_felem out[2],
+                                          size_t idx,
+                                          const fiat_p384_felem table[][2],
+                                          size_t table_size) {
+  OPENSSL_memset(out, 0, sizeof(fiat_p384_felem) * 2);
+  for (size_t i = 0; i < table_size; i++) {
+    fiat_p384_limb_t mismatch = i ^ idx;
+    fiat_p384_cmovznz(out[0], mismatch, table[i][0], out[0]);
+    fiat_p384_cmovznz(out[1], mismatch, table[i][1], out[1]);
+  }
+}
+
+// Multiplication of a point by a scalar, r = [scalar]P.
```

Comment:
This algorithm is left-to-right over the recoded scalar, so the first part of the analysis I wrote applies to it.